### PR TITLE
feat: add logger

### PR DESCRIPTION
**Pull request by beeme1mr**
Features include:

- A default logger that writes warning and errors to the console.
- A safe logger that will wrap user providers loggers to ensure stability
- Global logger can be updated with a custom logger
- Clients can be updated with a customer logger
- Custom client loggers override global loggers

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -246,9 +264,8 @@ export class OpenFeatureClient implements Client {
       try {
         await hook?.error?.(hookContext, err, options.hookHints);
       } catch (err) {
-        // TODO: replace with injected logger
-        console.error(`Unhandled error during 'error' hook: ${err}`);
-        console.error((err as Error).stack);
+        this.logger.error(`Unhandled error during 'error' hook: ${err}`);
+        this.logger.error((err as Error)?.stack);
```
Comment (toddbaert) [inline src/client.ts:268]: Good find here.

File: src/logger.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    try {
+      for (const level of LOG_LEVELS) {
+        if (!logger[level] || typeof logger[level] !== 'function') {
+          throw new Error(`The provided logger is missing the ${level} method.`);
+        }
+      }
+      this.logger = logger;
+    } catch (err) {
+      console.error(err);
+      console.warn('Falling back to the default logger.');
```
Comment (toddbaert) [inline src/logger.ts:30]: I fee like these should either both be `warn` or `error`.

File: src/logger.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  info(...args: unknown[]): void {
+    this.log('info', ...args);
+  }
+  debug(...args: unknown[]): void {
+    this.log('debug', ...args);
+  }
+
+  private log(level: keyof Logger, ...args: unknown[]) {
+    try {
+      this.logger[level](...args);
```
Comment (toddbaert) [inline src/logger.ts:50]: I actually think you  should call this with `.apply` or `.call` and pass the logger as the `thisArg`. If the logger method called any members, they would be undefined.

File: test/client.spec.ts [hunk 4]
```diff
@@ -138,51 +138,51 @@ describe(OpenFeatureClient.name, () => {
...
           const booleanDetails = await client.getBooleanDetails(booleanFlag, defaultBooleanValue);
 
           expect(booleanDetails.value).toEqual(BOOLEAN_VALUE);
           expect(booleanDetails.variant).toEqual(BOOLEAN_VARIANT);
-          expect(MOCK_PROVIDER.resolveBooleanEvaluation).toHaveBeenCalledWith(booleanFlag, defaultBooleanValue, {});
+          expect(MOCK_PROVIDER.resolveBooleanEvaluation).toHaveBeenCalledWith(booleanFlag, defaultBooleanValue, {}, {});
         });
       });
 
-      describe(OpenFeatureClient.prototype.getStringDetails.name, () => {
```
Comment (toddbaert) [inline test/client.spec.ts:153]: This makes me sad. These tests are all brittle now.

I'm not really a fan of changing these just for particular tooling.

File: src/logger.ts (hunk 3 above)
Comment (toddbaert) [reply src/logger.ts:50]: Nevermind, I think this is fine since you're still calling this as a method, not a function ref.

**Review by toddbaert — APPROVED** [review]

**Comment by toddbaert** [comment]
Looks good to me, good testing.

File: test/logger.spec.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+      expect(mockedLogger.info.mock.calls).toEqual([
+        ['in before hook'],
+        ['resolving boolean value'],
+        ['in after hook'],
+        ['in finally hook'],
+      ]);
+    });
+
+    it('should use the default logger in the error hook', async () => {
```
Comment (toddbaert) [inline test/logger.spec.ts:158]: Is this test name correct? It seems like were not using the default logger.

File: test/logger.spec.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+      expect(client.logger).toBeInstanceOf(DefaultLogger);
+    });
+
+    it('should provide a logger to the provider', async () => {
+      const client = OpenFeature.getClient();
+      const mockedLogger = new MockedLogger();
+      client.logger = mockedLogger;
+      await client.getBooleanValue('test', false);
+      expect(mockedLogger.info).toHaveBeenCalledWith('resolving boolean value');
```
Comment (toddbaert) [inline test/logger.spec.ts:140]: I think a lot of duplicated strings in assertions could be constants.

File: test/logger.spec.ts (hunk 5 above)
Comment (beeme1mr) [reply test/logger.spec.ts]: Good catch.

File: test/logger.spec.ts (hunk 6 above)
Comment (beeme1mr) [reply test/logger.spec.ts]: Done

File: src/client.ts (hunk 1 above)
Comment (dyladan) [inline src/client.ts:268]: Technically any type can be caught. A `typeof` check would avoid this casting

File: src/logger.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  warn(...args: unknown[]): void {
+    this.log('warn', ...args);
+  }
+  info(...args: unknown[]): void {
+    this.log('info', ...args);
+  }
+  debug(...args: unknown[]): void {
+    this.log('debug', ...args);
+  }
+
```
Comment (dyladan) [inline src/logger.ts:47]: These can be defined in a loop by your constructor and would make your output smaller. Not sure if minification is a priority for you.

Comment (beeme1mr) [reply src/logger.ts:47]: That's an interesting idea. I'll keep that in mind for future optimizations.

**Pull request merged**